test/integration/logeater_test.rb

```diff
@@ -60,7 +60,24 @@ class LogeaterTest < ActiveSupport::TestCase
 
     should "create an entry in the database" do
       assert_difference "Logeater::Request.count", +1 do
-        reader.import
+        logfile_reader.import
+      end
+    end
+  end
+
+
+  context "Given an app and a timestamp, import_since" do
+    setup do
+      log_sample = File.open(File.expand_path("./test/data/single_request.log"))
+      log_sample.lines do |line|
+        Logeater::Event.create(ep_app: app, original: line)
+      end
+      @events = Logeater::Event.all
+    end
+
+    should "import events since that given timestamp buuga" do
+      assert_difference "Logeater::Request.count", +1 do
+        eventfile_reader.import
```

Comment:
Should you also test that if you call `eventfile_reader.import` a second time, it doesn't create a duplicate entry?

Reply:
You could test the second scenario I wrote about like this:

```ruby
lines = File.open(File.expand_path("./test/data/single_request.log")).lines

# The first two lines will not be enough to describe a complete request
# so Logeater will not be able to create a request from them...
lines[0...2].each do |line|
  Logeater::Event.create(ep_app: app, original: line)
end
assert_no_difference "Logeater::Request.count" do
  eventfile_reader.import
end

# ...but if we later discover the rest of the lines that describe a complete
# request, it'd be good if Logeater could then recognize and import it.
lines[2..-1].each do |line|
  Logeater::Event.create(ep_app: app, original: line)
end
assert_difference "Logeater::Request.count", +1 do
  eventfile_reader.import
end
```

* * *

The question is: how to go back and catch those lines... 🤔 If we just back the timestamp up, we'll probably get duplicates, right? Or....... will Logeater know that it's already imported a given Request ID and skip it?

We _could_ put a `processed` flag on the `events` table and figure out how to set it when a corresponding Request gets inserted...

Reply:
Maybe surprisingly, this Just Works™. The test passed without needing to do anything further.

Reply:
👏